File: Source/Installer/ArchiveUtil.m
/* 
 *  ArchiveUtil.m
 *  
 *  Copyright 2021-2022 vChewing Project (3-Clause BSD License).
 *  Derived from 2011-2022 OpenVanilla Project (MIT License).
 *  Some rights reserved. See "LICENSE.TXT" for details.
 */

#import "ArchiveUtil.h"

@implementation ArchiveUtil
- (instancetype)initWithAppName:(NSString *)name
            targetAppBundleName:(NSString *)targetAppBundleName {
    self = [super init];
    if (self) {
        _appName = name;
        _targetAppBundleName = targetAppBundleName;
    }
    return self;
}

- (void)delloc {
    _appName = nil;
    _targetAppBundleName = nil;
}

- (BOOL)validateIfNotarizedArchiveExists {
    NSString *resourePath = [[NSBundle mainBundle] resourcePath];
    NSString *devModeAppBundlePath =
        [resourePath stringByAppendingPathComponent:_targetAppBundleName];

    NSArray<NSString *> *notarizedArchivesContent =
        [[NSFileManager defaultManager] subpathsAtPath:[self notarizedArchivesPath]];
    NSInteger count = [notarizedArchivesContent count];
    BOOL notarizedArchiveExists =
        [[NSFileManager defaultManager] fileExistsAtPath:[self notarizedArchive]];
    BOOL devModeAppBundleExists =
        [[NSFileManager defaultManager] fileExistsAtPath:devModeAppBundlePath];

    if (count > 0) {
        // Not a valid distribution package.
        if (count != 1 || !notarizedArchiveExists || devModeAppBundleExists) {
            NSAlert *alert = [[NSAlert alloc] init];
            [alert setAlertStyle:NSAlertStyleInformational];
            [alert setMessageText:@"Internal Error"];
            [alert
                setInformativeText:
                    [NSString stringWithFormat:@"devMode installer, expected archive name: %@, "
                                               @"archive exists: %d, devMode app bundle exists: %d",
                                               [self notarizedArchive], notarizedArchiveExists,
                                               devModeAppBundleExists]];
            [alert addButtonWithTitle:@"Terminate"];
            [alert runModal];

            [[NSApplication sharedApplication] terminate:nil];
        } else {
            return YES;
        }
    }

    if (!devModeAppBundleExists) {
        NSAlert *alert = [[NSAlert alloc] init];
        [alert setAlertStyle:NSAlertStyleInformational];
        [alert setMessageText:@"Internal Error"];
        [alert
            setInformativeText:[NSString stringWithFormat:@"Dev target bundle does not exist: %@",
                                                          devModeAppBundlePath]];
        [alert addButtonWithTitle:@"Terminate"];
        [alert runModal];
        [[NSApplication sharedApplication] terminate:nil];
    }

    // Notarized archive does not exist, but it's ok.
    return NO;
}

- (NSString *)unzipNotarizedArchive {
    if (![self validateIfNotarizedArchiveExists]) {
        return nil;
    }

    NSString *tempFilePath =
        [NSTemporaryDirectory() stringByAppendingPathComponent:[[NSUUID UUID] UUIDString]];
    NSArray *arguments = @[ [self notarizedArchive], @"-d", tempFilePath ];

    NSTask *unzipTask = [[NSTask alloc] init];
    [unzipTask setLaunchPath:@"/usr/bin/unzip"];
    [unzipTask setCurrentDirectoryPath:[[NSBundle mainBundle] resourcePath]];
    [unzipTask setArguments:arguments];
    [unzipTask launch];
    [unzipTask waitUntilExit];

    NSAssert(unzipTask.terminationStatus == 0, @"Must successfully unzipped");

    NSString *result = [tempFilePath stringByAppendingPathComponent:_targetAppBundleName];
    NSAssert([[NSFileManager defaultManager] fileExistsAtPath:result],
             @"App bundle must be unzipped at %@", result);
    return result;
}

- (NSString *)notarizedArchivesPath {
    NSString *resourePath = [[NSBundle mainBundle] resourcePath];
    NSString *notarizedArchivesPath =
        [resourePath stringByAppendingPathComponent:@"NotarizedArchives"];
    return notarizedArchivesPath;
}

- (NSString *)notarizedArchive {
    NSString *bundleVersion =
        [[[NSBundle mainBundle] infoDictionary] objectForKey:(id)kCFBundleVersionKey];
    NSString *notarizedArchiveBasename =
        [NSString stringWithFormat:@"%@-r%@.zip", _appName, bundleVersion];
    NSString *notarizedArchive =
        [[self notarizedArchivesPath] stringByAppendingPathComponent:notarizedArchiveBasename];
    return notarizedArchive;
}
@end

File: Installer/ArchiveUtil.swift
// Copyright (c) 2011 and onwards The OpenVanilla Project (MIT License).
// All possible vChewing-specific modifications are of:
// (c) 2021 and onwards The vChewing Project (MIT-NTL License).
// ====================
// This code is released under the MIT license (SPDX-License-Identifier: MIT)
// ... with NTL restriction stating that:
// No trademark license is granted to use the trade names, trademarks, service
// marks, or product names of Contributor, except as required to fulfill notice
// requirements defined in MIT License.

import Cocoa

struct ArchiveUtil {
  var appName: String
  var targetAppBundleName: String

  init(appName: String, targetAppBundleName: String) {
    self.appName = appName
    self.targetAppBundleName = targetAppBundleName
  }

  // Returns YES if (1) a zip file under
  // Resources/NotarizedArchives/$_appName-$bundleVersion.zip exists, and (2) if
  // Resources/$_invalidAppBundleName does not exist.
  func validateIfNotarizedArchiveExists() -> Bool {
    guard let resourePath = Bundle.main.resourcePath,
      let notarizedArchivesPath = notarizedArchivesPath,
      let notarizedArchive = notarizedArchive,
      let notarizedArchivesContent: [String] = try? FileManager.default.subpathsOfDirectory(
        atPath: notarizedArchivesPath)
    else {
      return false
    }

    let devModeAppBundlePath = (resourePath as NSString).appendingPathComponent(targetAppBundleName)
    let count = notarizedArchivesContent.count
    let notarizedArchiveExists = FileManager.default.fileExists(atPath: notarizedArchive)
    let devModeAppBundleExists = FileManager.default.fileExists(atPath: devModeAppBundlePath)

    if !notarizedArchivesContent.isEmpty {
      // 這裡不用「count > 0」，因為該整數變數只要「!isEmpty」那就必定滿足這個條件。
      if count != 1 || !notarizedArchiveExists || devModeAppBundleExists {
        let alert = NSAlert()
        alert.alertStyle = .informational
        alert.messageText = "Internal Error"
        alert.informativeText =
          "devMode installer, expected archive name: \(notarizedArchive), "
          + "archive exists: \(notarizedArchiveExists), devMode app bundle exists: \(devModeAppBundleExists)"
        alert.addButton(withTitle: "Terminate")
        alert.runModal()
        NSApp.terminate(nil)
      } else {
        return true
      }
    }

    if !devModeAppBundleExists {
      let alert = NSAlert()
      alert.alertStyle = .informational
      alert.messageText = "Internal Error"
      alert.informativeText = "Dev target bundle does not exist: \(devModeAppBundlePath)"
      alert.addButton(withTitle: "Terminate")
      alert.runModal()
      NSApp.terminate(nil)
    }

    return false
  }

  func unzipNotarizedArchive() -> String? {
    if !validateIfNotarizedArchiveExists() {
      return nil
    }
    guard let notarizedArchive = notarizedArchive,
      let resourcePath = Bundle.main.resourcePath
    else {
      return nil
    }
    let tempFilePath = (NSTemporaryDirectory() as NSString).appendingPathComponent(
      UUID().uuidString)
    let arguments: [String] = [notarizedArchive, "-d", tempFilePath]
    let unzipTask = Process()
    unzipTask.launchPath = "/usr/bin/unzip"
    unzipTask.currentDirectoryPath = resourcePath
    unzipTask.arguments = arguments
    unzipTask.launch()
    unzipTask.waitUntilExit()

    assert(unzipTask.terminationStatus == 0, "Must successfully unzipped")
    let result = (tempFilePath as NSString).appendingPathComponent(targetAppBundleName)
    assert(
      FileManager.default.fileExists(atPath: result),
      "App bundle must be unzipped at \(result)."
    )
    return result
  }

  private var notarizedArchivesPath: String? {
    guard let resourePath = Bundle.main.resourcePath else {
      return nil
    }
    let notarizedArchivesPath = (resourePath as NSString).appendingPathComponent(
      "NotarizedArchives")
    return notarizedArchivesPath
  }

  private var notarizedArchive: String? {
    guard let notarizedArchivesPath = notarizedArchivesPath,
      let bundleVersion = Bundle.main.infoDictionary?[kCFBundleVersionKey as String]
        as? String
    else {
      return nil
    }
    let notarizedArchiveBasename = "\(appName)-r\(bundleVersion).zip"
    let notarizedArchive = (notarizedArchivesPath as NSString).appendingPathComponent(
      notarizedArchiveBasename)
    return notarizedArchive
  }
}
